sample/src/main/kotlin/com/uber/autodispose/recipes/AutoDisposeFragmentKotlin.kt

```diff
@@ -0,0 +1,120 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.uber.autodispose.recipes
+
+import android.app.Fragment
+import android.content.Context
+import android.os.Bundle
+import android.view.View
+import com.uber.autodispose.LifecycleEndedException
+import com.uber.autodispose.LifecycleScopeProvider
+import io.reactivex.Observable
+import io.reactivex.functions.Function
+import io.reactivex.subjects.BehaviorSubject
+
+/**
+ * A [Fragment] example implementation for making one implement [LifecycleScopeProvider]. One would
+ * normally use this as a base fragment class to extend others from.
+ */
+abstract class AutoDisposeFragmentKotlin : Fragment(), LifecycleScopeProvider<AutoDisposeFragmentKotlin.FragmentEvent> {
+
+  private val lifecycleEvents = BehaviorSubject.create<FragmentEvent>()
+
+  enum class FragmentEvent {
+    ATTACH, CREATE, CREATE_VIEW, START, RESUME, PAUSE, STOP, DESTROY_VIEW, DESTROY, DETACH
+  }
+
+  override fun lifecycle(): Observable<FragmentEvent> {
+    return lifecycleEvents.hide()
+  }
+
+  override fun correspondingEvents(): Function<FragmentEvent, FragmentEvent> {
+    return CORRESPONDING_EVENTS
+  }
+
+  override fun peekLifecycle(): FragmentEvent? {
+    return lifecycleEvents.value
+  }
+
+  override fun onAttach(context: Context) {
+    super.onAttach(context)
+    lifecycleEvents.onNext(FragmentEvent.ATTACH)
+  }
+
+  override fun onCreate(savedInstanceState: Bundle) {
+    super.onCreate(savedInstanceState)
+    lifecycleEvents.onNext(FragmentEvent.CREATE)
+  }
+
+  override fun onViewCreated(view: View, savedInstanceState: Bundle) {
+    super.onViewCreated(view, savedInstanceState)
+    lifecycleEvents.onNext(FragmentEvent.CREATE_VIEW)
+  }
+
+  override fun onStart() {
+    super.onStart()
+    lifecycleEvents.onNext(FragmentEvent.START)
+  }
+
+  override fun onResume() {
+    super.onResume()
+    lifecycleEvents.onNext(FragmentEvent.RESUME)
+  }
+
+  override fun onPause() {
+    lifecycleEvents.onNext(FragmentEvent.PAUSE)
+    super.onPause()
+  }
+
+  override fun onStop() {
+    lifecycleEvents.onNext(FragmentEvent.STOP)
+    super.onStop()
+  }
+
+  override fun onDestroyView() {
+    lifecycleEvents.onNext(FragmentEvent.DESTROY_VIEW)
+    super.onDestroyView()
+  }
+
+  override fun onDestroy() {
+    lifecycleEvents.onNext(FragmentEvent.DESTROY)
+    super.onDestroy()
+  }
+
+  override fun onDetach() {
+    lifecycleEvents.onNext(FragmentEvent.DETACH)
+    super.onDetach()
+  }
+
+  companion object {
+
+    private val CORRESPONDING_EVENTS = Function<FragmentEvent, FragmentEvent> { event ->
+      when (event) {
+        AutoDisposeFragmentKotlin.FragmentEvent.ATTACH -> FragmentEvent.DETACH
```

Comment:
Both left and right-hand sides are of the same type but have different depth of name qualifiers.

Reply:
7b717f3